app/scripts/controllers/app-state-controller.ts

```diff
@@ -0,0 +1,869 @@
+    this.#onInactiveTimeout = onInactiveTimeout || (() => undefined);
+    this.store = new ObservableStore(
+      getDefaultAppStateControllerState(initState),
+    );
+    this.#timer = null;
+
+    this.isUnlocked = isUnlocked;
+    this.#waitingForUnlock = [];
+    addUnlockListener(this.handleUnlock.bind(this));
+
+    messenger.subscribe(
+      'PreferencesController:stateChange',
+      ({ preferences }: { preferences: Partial<Preferences> }) => {
+        const currentState = this.store.getState();
+        if (
+          typeof preferences?.autoLockTimeLimit === 'number' &&
+          currentState.timeoutMinutes !== preferences.autoLockTimeLimit
+        ) {
+          this._setInactiveTimeout(preferences.autoLockTimeLimit);
+        }
+      },
+    );
+
+    messenger.subscribe(
+      'KeyringController:qrKeyringStateChange',
+      (qrHardware: Json) =>
+        this.store.updateState({
+          qrHardware,
+        }),
+    );
+
+    const { preferences } = messenger.call('PreferencesController:getState');
+    if (typeof preferences.autoLockTimeLimit === 'number') {
+      this._setInactiveTimeout(preferences.autoLockTimeLimit);
+    }
+
+    this.#messagingSystem = messenger;
+    this.#approvalRequestId = null;
+  }
+
+  /**
+   * Get a Promise that resolves when the extension is unlocked.
+   * This Promise will never reject.
+   *
+   * @param shouldShowUnlockRequest - Whether the extension notification
+   * popup should be opened.
+   * @returns A promise that resolves when the extension is
+   * unlocked, or immediately if the extension is already unlocked.
+   */
+  getUnlockPromise(shouldShowUnlockRequest: boolean): Promise<void> {
+    return new Promise((resolve) => {
+      if (this.isUnlocked()) {
+        resolve();
+      } else {
+        this.waitForUnlock(resolve, shouldShowUnlockRequest);
+      }
+    });
+  }
+
+  /**
+   * Adds a Promise's resolve function to the waitingForUnlock queue.
+   * Also opens the extension popup if specified.
+   *
+   * @param resolve - A Promise's resolve function that will
+   * be called when the extension is unlocked.
+   * @param shouldShowUnlockRequest - Whether the extension notification
+   * popup should be opened.
+   */
+  waitForUnlock(resolve: () => void, shouldShowUnlockRequest: boolean): void {
+    this.#waitingForUnlock.push({ resolve });
+    this.emit(METAMASK_CONTROLLER_EVENTS.UPDATE_BADGE);
+    if (shouldShowUnlockRequest) {
+      this._requestApproval();
+    }
+  }
+
+  /**
+   * Drains the waitingForUnlock queue, resolving all the related Promises.
+   */
+  handleUnlock(): void {
+    if (this.#waitingForUnlock.length > 0) {
+      while (this.#waitingForUnlock.length > 0) {
+        this.#waitingForUnlock.shift()?.resolve();
+      }
+      this.emit(METAMASK_CONTROLLER_EVENTS.UPDATE_BADGE);
+    }
+
+    this._acceptApproval();
+  }
+
+  /**
+   * Sets the default home tab
+   *
+   * @param [defaultHomeActiveTabName] - the tab name
```

Comment:
Nit: Since we’ve defined the `defaultHomeActiveTabName` type, we can remove the square brackets `[]` from the JSDoc that marks the param as optional.
```suggestion
   * @param defaultHomeActiveTabName - the tab name
```